src/exclusion/object_filter.hpp

```diff
@@ -22,35 +24,131 @@
 namespace ddwaf::exclusion {
 
 class path_trie {
-public:
-    path_trie() = default;
-
-    [[nodiscard]] path_trie find(std::string_view key) const;
-
-    template <typename T> [[nodiscard]] path_trie find(const std::vector<T> &path) const;
-
-    template <typename T> void insert(const std::vector<T> &path);
-
-    [[nodiscard]] bool is_terminal() const { return root && root->terminal; }
-    [[nodiscard]] bool is_valid() const { return root && !root->values.empty(); }
-
-protected:
+private:
     struct trie_node {
-        // Heterogenous lookups required...
-        std::unordered_map<std::string_view, std::shared_ptr<trie_node>> values;
-        // TODO this shouldn't be necessary
-        bool terminal{false};
+        std::string path_component;
+        std::vector<trie_node> children{};
+
+        explicit trie_node(std::string path_component) :
+            path_component{std::move(path_component)} {}
+
+        ~trie_node() = default;
+        trie_node(const trie_node&) = delete;
+        trie_node(trie_node&&) = default;
+        trie_node& operator=(const trie_node&) = delete;
+        trie_node& operator=(trie_node&&) = default;
+
+        struct comp {
+            bool operator()(const trie_node& n, std::string_view v) {
+                return std::string_view{n.path_component} < v;
+            }
+            bool operator()(std::string_view v, const trie_node& n) {
+                return v < std::string_view{n.path_component};
+            }
+        };
+
+        [[nodiscard]] trie_node const *get_child(std::string_view key) const {
+            auto [beg, end] = std::equal_range(
+                children.begin(), children.end(), key, comp{});
+            if (beg != end) {
+                return &*beg;
+            }
+            return nullptr;
+        }
+
+        trie_node& get_or_create_child(std::string_view key, bool& is_new) {
+            auto [beg, end] = std::equal_range(
+                children.begin(), children.end(), key, comp{});
+            if (beg != end) {
+                is_new = false;
+                return *beg;
+            }
+
+            auto ins_it = children.emplace(end, std::string{key});
+            is_new = true;
+            return *ins_it;
+        }
+
+        [[nodiscard]] bool is_terminal() const {
+            return children.empty();
+        }
     };
+    static_assert(std::is_move_assignable<trie_node>::value);
 
-    explicit path_trie(std::shared_ptr<trie_node> node_) : root(std::move(node_)) {}
+    trie_node root{""};
+    // we allow adding the root to the trie (matching everything)
+    // so we use this flag to distinguish the two cases (empty vs everything)
+    bool is_empty{true}; 
 
-    std::string_view get_stored_string(std::string_view str);
+public:
+    class traverser {
+        trie_node const *const cur_node{};
+    public:    
+        enum class state {
+            NOT_FOUND,
+            FOUND,
+            GLUE
```

Comment:
nit: I don't usually use capital letters for enum class values since they are nicely namespaced and type-safe so it's already clear they are "constants".

Reply:
Also I'm unconvinced about `GLUE`, it's unclear what that's supposed to mean, although I don't have good suggestions without referring to trie internals.

Reply:
I'll rename it to something `INTERMEDIATE_NODE`